tokenawarebandit.go

```diff
@@ -0,0 +1,614 @@
+}
+
+func (t *TokenAwareBanditHostPolicy) KeyspaceChanged(update KeyspaceUpdateEvent) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.updateReplicas(update.Keyspace)
+}
+
+// updateReplicas initiates updating of replicas for the given keyspace.
+// It must be called while the caller holds t.mu mutex locked for writes.
+func (t *TokenAwareBanditHostPolicy) updateReplicas(keyspace string) {
+	ks, ok := t.keyspaces[keyspace]
+	if !ok {
+		ks = &tokenAwareBanditKeyspace{}
+		t.keyspaces[keyspace] = ks
+	}
+
+	ks.needsUpdate = true
+
+	if ks.updating {
+		return
+	}
+
+	ks.updating = true
+	go t.updateReplicasLoop(ks, keyspace)
+}
+
+// replicasNeedUpdate checks if the keyspace needs updating.
+// If the keyspace does not need to be updated, it resets the updating flag.
+func (t *TokenAwareBanditHostPolicy) replicasNeedUpdate(ks *tokenAwareBanditKeyspace) bool {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if ks.needsUpdate {
+		ks.needsUpdate = false
+		return true
+	}
+	ks.updating = false
+	return false
+}
+
+// updateReplicasLoop performs actual update of replicas for a keyspace.
+// updateReplicasLoop must be called in a goroutine not holding the t.mu mutex.
+func (t *TokenAwareBanditHostPolicy) updateReplicasLoop(ks *tokenAwareBanditKeyspace, keyspaceName string) {
+	for t.replicasNeedUpdate(ks) {
+		// We can't call getKeyspaceMetadata while holding the lock since getKeyspaceMetadata can do I/O.
+		metadata, err := t.getKeyspaceMetadata(keyspaceName)
+		if err != nil {
+			t.logger.Println(err)
+			continue
+		}
+
+		strategy := getStrategy(metadata)
+		var replicas tokenRingReplicas
+
+		if strategy != nil {
+			t.mu.RLock()
+			tokenRing := t.tokenRing
+			t.mu.RUnlock()
+			replicas = strategy.replicaMap(tokenRing)
+		}
+
+		t.mu.Lock()
+		ks.replicas = replicas
+		t.mu.Unlock()
+	}
+	if t.afterReplicasUpdated != nil {
+		t.afterReplicasUpdated(keyspaceName)
+	}
+}
+
+// SetPartitioner is called to set the partitioner used.
+func (t *TokenAwareBanditHostPolicy) SetPartitioner(partitioner string) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if t.partitioner == partitioner {
+		return
+	}
+
+	t.partitioner = partitioner
+
+	err := t.resetRing()
+	if err != nil {
+		Logger.Println(err)
+	}
+}
+
+// resetRing replaces the token ring with a new one.
+// must be called with t.mu lock held for writing.
+func (t *TokenAwareBanditHostPolicy) resetRing() error {
+	newTR, err := newTokenRing(t.partitioner, t.hostInfoSlice())
+	if err != nil {
+		return err
+	}
+	t.tokenRing = newTR
+
+	for keyspaceName := range t.keyspaces {
+		t.updateReplicas(keyspaceName)
+	}
+
+	return nil
+}
+
+// AddHost adds a single host to the policy after it was discovered/added to the cluster.
+func (t *TokenAwareBanditHostPolicy) AddHost(host *HostInfo) {
+	t.AddHosts([]*HostInfo{host})
+}
+
+// AddHosts adds multiple hosts to the policy after they were discovered/added to the cluster.
+func (t *TokenAwareBanditHostPolicy) AddHosts(hosts []*HostInfo) {
```

Comment:
Can we unify two methods and use a variadic parameter instead?

```suggestion
// AddHosts adds multiple hosts to the policy after they were discovered/added to the cluster.
func (t *TokenAwareBanditHostPolicy) AddHosts(hosts ...*HostInfo) {
```

Reply:
No, we need to implement HostSelectionPolicy https://github.com/kiwicom/gocql/blob/bd3bbd7282a06687131e84a77320c132005d9db1/policies.go#L299-L312
resp. HostStateNotifier
https://github.com/kiwicom/gocql/blob/bd3bbd7282a06687131e84a77320c132005d9db1/policies.go#L287-L292
so we can't change the signature of AddHost.

I'll add a comment explaining this.